### PR TITLE
docs: Migrate datatypes theory reference.

### PR DESCRIPTION
**Pull request by aniemetz**
This migrates https://cvc4.github.io/datatypes.

### CODE REVIEW

**Comment by aniemetz** [comment]
@ajreynol please check in particular if the C++ API examples in the table are correct.

**Comment by aniemetz** [comment]
You'll find the page here when it is done building: https://cvc5.github.io/docs-ci/docs-pr6662

File: docs/conf.py.in [hunk 1]
```diff
@@ -74,3 +74,73 @@ cpp_index_common_prefix = ["cvc5::api::"]
 
 # where to look for include-build-file
 ibf_folders = ['${CMAKE_CURRENT_BINARY_DIR}']
+
+# -- SMT-LIB syntax highlighting ---------------------------------------------
+from pygments.lexer import RegexLexer
+from pygments import token
```
Comment (nafur) [inline docs/conf.py.in:80]: I'm glad you added this lexer! However, given that the list of tokens will probably grow quite a bit, can we move it to a separate file?

File: docs/theories/datatypes.rst [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+   (((Red) (Black))))
+
+A List of Int with ``cons`` and ``nil`` as constructors:::
+
+.. code:: smtlib
+
+  (declare-datatypes ((list 0))
+   (((cons (head Int) (tail list)) (nil))))
+
+A parametric List of T's:::
```
Comment (nafur) [inline docs/theories/datatypes.rst:68]: The triple colons seem to mess with the Sphinx markup...
```suggestion
A List of Int with ``cons`` and ``nil`` as constructors:

.. code:: smtlib

  (declare-datatypes ((list 0))
   (((cons (head Int) (tail list)) (nil))))

A parametric List of T's:
```

File: docs/conf.py.in [hunk 3]
```diff
@@ -74,3 +74,73 @@ cpp_index_common_prefix = ["cvc5::api::"]
...
+
+    tokens = {
+        'root': [
+            (r'QF_BV', token.Text),
+            (r'QF_UFDT', token.Text),
+            (r'ALL_SUPPORTED', token.Text),
+            (r'set-logic', token.Keyword),
+            (r'set-option', token.Keyword),
+            (r'declare-const', token.Keyword),
+            (r'declare-datatypes', token.Keyword),
```
Comment (nafur) [inline docs/conf.py.in:94]:
```suggestion
            (r'declare-datatypes', token.Keyword),
            (r'declare-codatatypes', token.Keyword),
```

File: docs/conf.py.in (hunk 1 above)
Comment (aniemetz) [reply docs/conf.py.in:80]: This is not worth the headache in my opinion. Requires copying the file into the build directory and so on.

Comment (nafur) [reply docs/conf.py.in:80]: We could put it into the `ext` directory, which is put into the path anyway.

File: docs/theories/datatypes.rst (hunk 2 above)
Comment (aniemetz) [reply docs/theories/datatypes.rst:68]: Yeah, that was unintended. Thanks for catching.

File: docs/conf.py.in (hunk 1 above)
Comment (aniemetz) [reply docs/conf.py.in]: Oh nice, that works! Updated.

**Review by nafur — APPROVED** [review]
LGTM.

**Pull request merged**